rust/core-lib/src/view.rs

```diff
@@ -790,51 +805,112 @@ impl View {
                     }
                 }
                 RenderTactic::Render => {
-                    // TODO: update (rather than re-render) in cases of text valid
-                    if seg.validity == line_cache_shadow::ALL_VALID {
+                    // Depending on the state of TEXT_VALID, STYLES_VALID and
+                    // CURSOR_VALID, perform one of the following actions:
+                    //
+                    //   - All three are valid => send the "copy" op
+                    //     (+preceding "skip" to catch up with "ln" to update);
+                    //
+                    //   - Text and styles are valid, cursors are not => same,
+                    //     but send an "update" op instead of "copy" to move
+                    //     the cursors;
+                    //
+                    //   - Text or styles are invalid => re-encode the lines
+                    //     in JSON and send the new content in an "insert" op.
+                    if (seg.validity & line_cache_shadow::TEXT_VALID) != 0
+                        && (seg.validity & line_cache_shadow::STYLES_VALID) != 0
+                    {
                         let n_skip = seg.their_line_num - line_num;
                         if n_skip > 0 {
                             ops.push(UpdateOp::skip(n_skip));
                         }
                         let line_offset = self.offset_of_line(text, seg.our_line_num);
-                        let logical_line = text.line_of_offset(line_offset) + 1;
-                        ops.push(UpdateOp::copy(seg.n, logical_line));
-                        b.add_span(seg.n, seg.our_line_num, line_cache_shadow::ALL_VALID);
+                        let logical_line = text.line_of_offset(line_offset);
+
+                        if (seg.validity & line_cache_shadow::CURSOR_VALID) != 0 {
+                            // ALL_VALID; copy lines as-is
+                            ops.push(UpdateOp::copy(seg.n, logical_line + 1));
+                        } else {
+                            // !CURSOR_VALID; update cursors
+                            let start_line = seg.our_line_num;
+
+                            let rendered_lines = self
+                                .lines
+                                .iter_lines(text, start_line)
+                                .take(seg.n)
+                                .map(|l| {
+                                    self.render_line(
+                                        client, styles, l, /* text = */ None,
+                                        /* style_spans = */ None,
+                                    )
+                                })
+                                .collect::<Vec<_>>();
+
+                            let logical_line_opt =
+                                if logical_line == 0 { None } else { Some(logical_line + 1) };
+                            ops.push(UpdateOp::update(rendered_lines, logical_line_opt));
```

Comment:
Isn't this the sme as on line 827? Could we factor this out into a seperate function or similiar so we don't have the same code twice?

Reply:
Correct. The same code duplication existed in the match arms before my change. This could have been done intentionally (for simplicity). But I will try to factorize it out if that doesn't make the call graph too complex.